cpp/src/arrow/util/formatting.h

```diff
@@ -0,0 +1,237 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// This is a private header for number-to-string formatting utilities
+
+#include <cassert>
+#include <limits>
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <utility>
+
+#include <double-conversion/double-conversion.h>
+
+#include "arrow/status.h"
+#include "arrow/type.h"
+#include "arrow/type_traits.h"
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/config.h"
+
+namespace arrow {
+namespace internal {
+
+/////////////////////////////////////////////////////////////////////////
+// Boolean formatting
+
+template <typename ARROW_TYPE, typename Enable = void>
+class StringFormatter;
+
+template <>
+class StringFormatter<BooleanType> {
+ public:
+  explicit StringFormatter(const std::shared_ptr<DataType>& = NULLPTR) {}
+
+  using value_type = bool;
+
+  template <typename Appender>
+  Status operator()(bool value, Appender&& append) {
+    if (value) {
+      const char string[] = "true";
+      return append(string, static_cast<int32_t>(sizeof(string) - 1));
+    } else {
+      const char string[] = "false";
+      return append(string, static_cast<int32_t>(sizeof(string) - 1));
+    }
+  }
+};
+
+/////////////////////////////////////////////////////////////////////////
+// Integer formatting
+
+namespace detail {
+
+// A 2x100 direct table mapping integers in [0..99] to their decimal representations.
+ARROW_EXPORT extern const char digit_pairs[];
```

Comment:
Interesting. Out of curiosity, how much performance does this buy you?

Reply:
It's around 30% slower if computing one digit at a time.